File: matlab/physiomon_settings.m
%
% physiomon_settings
%
%   reads the <physiomon> *.ini file and stores this information into 
%          - m_channels
%          - m_displays
%          - m_events
%
%   to avoid conflicts with the matlab <settings> this class is named <physiomon_settings>
%
% modifications
%   25-jan-2024   JM    initial version
%   11-feb-2024   JM    now derivated from a handle class (pointer logic)
%   28-feb-2024   JM    numeric display added

classdef physiomon_settings < handle

  properties (Access = private)
    
    m_filename = [];
    m_channels = [];
    m_displays = [];
    m_stores = [];
    m_events = [];

    defs = defines();       % load the defines

  end

  methods 
  
    %% constructor

    function obj = physiomon_settings()
    
      % constructor of the class, sets the filename. location is dependent of the
      % operating system

      if ismac(), obj.m_filename = '/Users/jan/.config/JanSoft/physiomon.ini'; end
      
      obj.m_events = repmat("",obj.defs.MAX_EVENTS,1);
    
    end

    %% getChannels

    function channels = getChannels(obj)
    
      % this method returns the private property <m_channels>
      %
      %     syntax : channels = getChannels(obj)
      %
      % with <channels> a copy of the private member m_channels

      channels = obj.m_channels;

    end

    %% getEvents

    function events = getEvents(obj)
    
      % this method returns the private property <m_events>
      %
      %     syntax : events = getEvents(obj)
      %
      % with <events> a copy of the private member m_events
      
      events = obj.m_events;

    end  

    %% getDisplays

    function displays = getDisplays(obj)
    
      % this method returns the private property <m_displays>
      %
      %     syntax : displays = getDisplays(obj)
      %
      % with <displays> a copy of the private member m_displays
      
      displays = obj.m_displays;

    end  

    %% setChannels

    function setChannels(obj,channels)
    
      % the method <setChannels> copies information from <channels> to the internal
      % private propertie <m_channels> This method is used in the dialog boxes
      %
      %     syntax : setChannels(obj,channels)
      %
      % with <channels> an vector of a channelStruct

      obj.m_channels = channels;

    end

    %% setDisplays

    function setDisplays(obj,displays)

      % the method <setDisplays> copies information from <displays> to the internal
      % private propertie <m_displays> This method is used in the dialog boxes
      %
      %     syntax : setDisplays(obj,displays)
      %
      % with <displays> an vector of a displayStruct

      obj.m_displays = displays;

    end

    %% setEvents

    function setEvents(obj,events)

      % the method <setEvents> copies an list of events (<events>) to the internal
      % private propertie <m_events> This method is used in the dialog boxes
      %
      %     syntax : setEvents(obj,events)
      %
      % with <events> a list of events
      
      obj.m_events = events;
    
    end

    %% iniRead

    function devicename = iniRead(obj)
    
      % reads the *.INI file and make 4 structures, depending on the information in the *.INI
      % file
      %     - m_channels    : containing information about channels
      %     - m_displays    : contains the event settings
      %     - m_stores      : contains store settings 
      %     - m_events      : contains information for the events
      %
      %   syntax : devicdename = iniRead(obj)
      %
      % this method differs from the one used in Qt (python and C++). No defaults values
      % are given except for the event structure. <devicename> is returned, the name of
      % the device (and so the device *.INI file)

      tmpStruct = ini2struct(obj.m_filename);
      
      numChan = round(str2double(tmpStruct.algemeen.numchan));
      numDisp = round(str2double(tmpStruct.algemeen.numdisp));
      numStore = round(str2double(tmpStruct.algemeen.numstore));

      obj.readDisplays(tmpStruct,numDisp);
      obj.readChannels(tmpStruct,numChan);
      obj.readStores(tmpStruct,numStore);
      obj.readEvents(tmpStruct,obj.defs.MAX_EVENTS);
      
      devicename = tmpStruct.algemeen.device;
    end

  end

  methods (Access = private)
  
    % readChannels

    function readChannels(obj,tmpStruct,numChannels)
      
      % readChannels decodes the <tmpStruct>, containing the whole *.INI file in an array
      % of <m_channels> with size <numChannels
      %
      %   syntax : readChannels(obj,tmpStruct,numChannels)
      %
      % <m_channels> is an internal private member of the class

      obj.m_channels = repmat(struct('name',[],'type',[],'source',[],'display',[],'precision',[],'color',[]),numChannels,1);

      for iChan = 1:numChannels

        section = ['channel' num2str(iChan)];

        obj.m_channels(iChan).name = tmpStruct.(section).name;
        obj.m_channels(iChan).source = str2double(tmpStruct.(section).source);
        obj.m_channels(iChan).display = str2double(tmpStruct.(section).display);
        obj.m_channels(iChan).precision = str2double(tmpStruct.(section).precision);

        switch lower(tmpStruct.(section).type) 

          case "analog in" 
            obj.m_channels(iChan).type = obj.defs.TYPE_ANALOG_IN;
        end

        switch lower(tmpStruct.(section).color)
          
          case "red"
            obj.m_channels(iChan).color = obj.defs.COLOR_RED;
          case "green"
            obj.m_channels(iChan).color = obj.defs.COLOR_GREEN;
          case "blue"
            obj.m_channels(iChan).color = obj.defs.COLOR_BLUE;
          case "cyan"
            obj.m_channels(iChan).color = obj.defs.COLOR_CYAN;
          case "magenta"
            obj.m_channels(iChan).color = obj.defs.COLOR_MAGENTA;
          case "yellow"
            obj.m_channels(iChan).color = obj.defs.COLOR_YELLOW;
          case "black"
            obj.m_channels(iChan).color = obj.defs.COLOR_BLACK;
          case "white"
            obj.m_channels(iChan).color = obj.defs.COLOR_WHITE;
          otherwise
            obj.m_channels(iChan).color = obj.defs.COLOR_GREEN;
        end

      end
    end

    % readDisplays

    function readDisplays(obj,tmpStruct,numDisplays)

      % readDisplay decodes the <tmpStruct>, containing the whole *.INI file in an array
      % of <m_displays> with size <numDisplays
      %
      %   syntax : readDisplays(obj,tmpStruct,numDisplays)
      %
      % <m_displays> is an internal private member of the class

      obj.m_displays = repmat(struct('top',[],'left',[],'width',[],'height',[],'ymin',[],'ymax',[],'timescale',[],'mode',[]),numDisplays,1);

      for iDisplay = 1:numDisplays
      
        section = ['display' num2str(iDisplay)];

        obj.m_displays(iDisplay).top = str2double(tmpStruct.(section).top);
        obj.m_displays(iDisplay).left = str2double(tmpStruct.(section).left);
        obj.m_displays(iDisplay).width = str2double(tmpStruct.(section).width);
        obj.m_displays(iDisplay).height = str2double(tmpStruct.(section).height);
        obj.m_displays(iDisplay).ymin = str2double(tmpStruct.(section).ymin);        
        obj.m_displays(iDisplay).ymax = str2double(tmpStruct.(section).ymax);        
        obj.m_displays(iDisplay).timescale = str2double(tmpStruct.(section).timescale);

        switch (lower(tmpStruct.(section).updatemode))
          
          case 'strip'
            obj.m_displays(iDisplay).mode = obj.defs.DISPLAY_MODE_STRIP;
          case 'sweep'
            obj.m_displays(iDisplay).mode = obj.defs.DISPLAY_MODE_SWEEP;
          case 'scope'
            obj.m_displays(iDisplay).mode = obj.defs.DISPLAY_MODE_SCOPE;
          case 'numeric'
            obj.m_displays(iDisplay).mode = obj.defs.DISPLAY_MODE_NUMERIC;
          otherwise
            error('unknown display mode');
        end

      end

    end
        
    % readStores

    function readStores(obj,tmpStruct,numStores)
    
      % readStores reads the groupname of the store engines
      %
      %   syntax : readStores(obj,tmpStruct,numStores)
      %
      % <m_stores> is an internal private member of the class. <tmpStruct> is the 
      % tempory structure created from reading the *.IN* file and <numStores> are the 
      % number of stores/group. 

      obj.m_stores = repmat(struct('groupname',[]),numStores,1);

      for iStore = 1:numStores

        section = ['store' num2str(iChan)];
        obj.m_channels(iStore).groupname = tmpStruct.(section).name;

      end
      
    end

    % readEvents

    function readEvents(obj,tmpStruct,numEvents)
    
      % readEvents decodes the <tmpStruct>, containing the whole *.INI file in an array
      % of <m_events> with size <numEvents>
      %
      %   syntax : readEvents(obj,tmpStruct,numEvents)
      %
      % <m_events> is an internal private member of the class. If the event number in the
      % *.INI file is not found a default string "this is event <nr>" is added

      for iEvent = 1:numEvents, obj.m_events(iEvent) = ['this is event ' num2str(iEvent)]; end

      for iEvent = 1:numEvents
        field = ['x' num2str(iEvent)];
        if isfield(tmpStruct.events,field), obj.m_events(iEvent) = tmpStruct.events.(field); end
      end
    end


  end
end

File: python/physiomon_settings.py
#
# physiomon_settings
#
#     class containing the settings for the program
#
#   modifications
#     18-dec-2023   JM    initial version
#     11-feb-2024   JM    renamed to <physiomon_settings>

TYPE_ANALOG_IN = 1

COLOR_RED = 0
COLOR_GREEN = 1
COLOR_BLUE = 2
COLOR_CYAN = 3
COLOR_MAGENTA = 4
COLOR_YELLOW = 5
COLOR_BLACK = 6
COLOR_WHITE = 7

from PySide6.QtCore import QSettings

class physiomon_settings() :

  # constructor

  def __init__(self) :
    
    self.m_settings = QSettings(QSettings.IniFormat,QSettings.UserScope,"JanSoft","physiomon")

    #-jm self.m_general = {"numchan" : 1, "numdisp" : 1, "device" :""}

    self.m_numchan = 0
    self.m_numdisp = 0
    self.m_numstore = 0
    self.m_device = ""

    self.m_channels = []
    self.m_displays = []
    self.m_stores = []
    self.m_events = []

    self.m_settings.setFallbacksEnabled(False)
    
    return

  # getEventString
  #
  #   returns the event string belonging to the event

  def getEventString(self,key) :
  
    return self.m_events[key] 
  
  # iniRead
  #
  #   read the *.INI file

  def iniRead(self) :

    self._readGeneral()
    self._readChannels()
    self._readDisplaySettings()
    self._readStoreSettings()
    self._readEventSettings()

  #-jm  return self.m_general["device"]
    return self.m_device
  
  # iniWrite
  #
  #     writes the current configuration

  def iniWrite(self) :

    self.m_settings.beginGroup("general")
    self.m_settings.setValue("numchan",self.m_numchan)
    self.m_settings.setValue("numdisp",self.m_numdisp)
                                         
    self.endGroup()


    return
  
  # _readGeneral
  #
  #     private method to read the general settings from the *.ini file

  def _readGeneral(self) :

    self.m_settings.beginGroup("algemeen")
    self.m_numchan = int(self.m_settings.value("numchan",defaultValue = 1))
    self.m_numdisp = int(self.m_settings.value("numdisp",1))
    self.m_numstore = int(self.m_settings.value("numstore",1))
    self.m_device = self.m_settings.value("device","")
    self.m_settings.endGroup()
    
    return
  
  # _readChannels
  #
  #     private method to read the settings for each channel

  def _readChannels(self) :

    channel = {"name" : "", "type" : 0, "source" : 0, "display" : 0, "precision" : 1, "color" : COLOR_GREEN}
    self.m_channels = []

    for i in range(self.m_numchan) :  
      keyName = "channel " + str(i+1)
      
      self.m_settings.beginGroup(keyName)
      channel["name"] = self.m_settings.value("name",1)
      channel["source"] = int(self.m_settings.value("source", defaultValue = 0))
      channel["display"] = int(self.m_settings.value("display",defaultValue = 0))
      channel["precision"] = int(self.m_settings.value("precision",defaultValue = 1))
      
      typeSignal = self.m_settings.value("type",defaultValue = "analog in")
      if (typeSignal == 'analog in') : channel["type"] = TYPE_ANALOG_IN
   
      color = self.m_settings.value("color", defaultValue = "green")
      if (color.lower() == "red") : channel["color"] = COLOR_RED
      if (color.lower() == "green") : channel["color"] = COLOR_GREEN
      if (color.lower() == "blue") : channel["color"] = COLOR_BLUE
      if (color.lower() == "cyan") : channel["color"] = COLOR_CYAN
      if (color.lower() == "magenta") : channel["color"] = COLOR_MAGENTA
      if (color.lower() == "yellow") : channel["color"] = COLOR_YELLOW
      if (color.lower() == "black") : channel["color"] = COLOR_BLACK
      if (color.lower() == "white") : channel["color"] = COLOR_WHITE
      
      self.m_settings.endGroup()
      
      self.m_channels.append(channel.copy())

    return

  # _readDisplaySettings
  #
  #     private method to read the display settings

  def _readDisplaySettings(self) :

    display = {"top" : 0, "left" : 0, "width" : 0, "height" : 0, "ymin" : 0, "ymax" : 0, "timescale" : 10, "mode" : ""}
    self.m_displays = []

    for i in range(self.m_numdisp) :

      name = "display " + str(i+1)

      self.m_settings.beginGroup(name)
      display["top"] = float(self.m_settings.value("top",0.0))
      display["left"] = float(self.m_settings.value("left",0.0))
      display["width"] = float(self.m_settings.value("width",0.0))
      display["height"] = float(self.m_settings.value("height",0.0))
      display["ymin"] = float(self.m_settings.value("ymin",0.0))
      display["ymax"] = float(self.m_settings.value("ymax",0.0))
      display["timescale"] = float(self.m_settings.value("timescale",10))
      display["mode"] = self.m_settings.value("update mode","scope")
      self.m_settings.endGroup()

      self.m_displays.append(display.copy())
    
    return  
  
  # readStoreSettings
  #
  #     read the store information (name of groups)

  def _readStoreSettings(self) :

    store = {"groupname" : ""}
    self.m_stores = []

    for i in range(self.m_numstore) :  
      
      keyName = "store " + str(i+1)
      
      self.m_settings.beginGroup(keyName)
      store["groupname"] = self.m_settings.value("groupname",1)
      self.m_settings.endGroup()
      
      self.m_stores.append(store.copy())

    return
  
  #  readEventSettings
  #
  #     private method to read the text for the events

  def _readEventSettings(self) :

    self.m_settings.beginGroup("events")
    for i in range(1,11) :
      event = self.m_settings.value(str(i),"event " + str(i))
      self.m_events.append(event)
    self.m_settings.endGroup()
    return
